### PR TITLE
Added SnakeGame

### PR DESCRIPTION
**Pull request by todorzag**

### CODE REVIEW

File: SnakeGame/SnakeGame/Game.cs [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+        {
+            bool hasWalls = AskHasWalls();
+            _gameBoard = new GameBoard(hasWalls);
+
+            SetUserName();
+
+            Console.Clear();
+            Console.WriteLine(Logos.GameStartLogo);
+
+            Thread.Sleep(2000);
```
Comment (todorzag) [inline SnakeGame/SnakeGame/Game.cs]: How to move Thread.Sleep() to main?

File: SnakeGame/SnakeGame/Game.cs [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+    {
+        enum Directions
+        {
+            Right,
+            Left,
+            Up,
+            Down
+        }
+
+        public delegate void Del();
```
Comment (todorzag) [inline SnakeGame/SnakeGame/Game.cs]: Need to remove

File: SnakeGame/SnakeGame/Snake.cs [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+        public (int, int) CurrentPosition
+        {
+            get => SnakeHead.Position;
+        }
+
+        public Snake()
+        {
+            SnakeHead = new SnakePart(1, 4, "○");
+
+            _snakeBody = new List<SnakePart>
```
Comment (mvlt) [inline SnakeGame/SnakeGame/Snake.cs]: What if we want to make starting length of the snake configurable

File: SnakeGame/SnakeGame/Snake.cs [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+        }
+
+        public (int, int) CurrentPosition
+        {
+            get => SnakeHead.Position;
+        }
+
+        public Snake()
+        {
+            SnakeHead = new SnakePart(1, 4, "○");
```
Comment (mvlt) [inline SnakeGame/SnakeGame/Snake.cs]: Isn't the head part of the body

File: SnakeGame/SnakeGame/SnakePart.cs [hunk 5]
```diff
@@ -0,0 +1,27 @@
+﻿namespace SnakeGame
+{
+    public class SnakePart
+    {
+        public int X { get; set; }
+        public int Y { get; set; }
+        public string Symbol { get; }
+        public (int, int) Position { get => (X, Y); }
+
+        public SnakePart(int x, int y)
```
Comment (mvlt) [inline SnakeGame/SnakeGame/SnakePart.cs]: Do we need 2 constructors for different type of displaying. Snake part should be able to print itself differently based on a param

File: SnakeGame/SnakeGame/Snake.cs [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+        }
+
+        public bool CheckSpawnOnSnake((int, int) applePosition)
+        {
+            return _snakeBody.Any(x => x.Position == applePosition);
+        }
+
+        public void MoveX(int x)
+        {
+            SnakeHead.X += x;
```
Comment (mvlt) [inline SnakeGame/SnakeGame/Snake.cs]: Why don't we check here if the move is appropriate?

File: SnakeGame/SnakeGame/Snake.cs [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+        }
+
+        public void MoveY(int y)
+        {
+            SnakeHead.Y += y;
+        }
+
+        public void UpdateBodyPosition()
```
Comment (mvlt) [inline SnakeGame/SnakeGame/Snake.cs]: Another way which we could try is adding a new element in front and removing the last one.

File: SnakeGame/SnakeGame/Snake.cs (hunk 6 above)
Comment (mvlt) [reply SnakeGame/SnakeGame/Snake.cs]: Also do the teleport here

File: SnakeGame/SnakeGame/Game.cs [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+            movementDirections = new Dictionary<string, Directions>
+            {
+                {"RightArrow", Directions.Right },
+                {"LeftArrow", Directions.Left },
+                {"DownArrow", Directions.Down },
+                {"UpArrow", Directions.Up }
+            };
+        }
+
+        private bool ShouldEndGame()
```
Comment (mvlt) [inline SnakeGame/SnakeGame/Game.cs]: Shouldn't this be part of the Snake class

File: SnakeGame/SnakeGame/Game.cs [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+        {
+            return movementDirections.ContainsKey(keyPressed);
+        }
+
+        private void MoveSnake(int directionNum, Action<int> moveMethod)
+        {
+            _snake.ClearLastSnakePart();
+            _snake.UpdateBodyPosition();
+
+            moveMethod(directionNum);
```
Comment (mvlt) [inline SnakeGame/SnakeGame/Game.cs]: Better use try catch and move the logic for teleport and game over checks in snake class

File: SnakeGame/SnakeGame/Snake.cs [hunk 10]
```diff
@@ -0,0 +1,156 @@
+﻿namespace SnakeGame
+{
+    public class Snake
+    {
+        public SnakePart SnakeHead;
```
Comment (mvlt) [inline SnakeGame/SnakeGame/Snake.cs]: In Snake class shouldn't be just named Head (Snake.Head not Snake.SnakeHead)

**Review by mvlt — CHANGES_REQUESTED** [review]

File: SnakeGame/SnakeGame/Game.cs [hunk 11]
```diff
@@ -0,0 +1,234 @@
...
+
+        public void Over()
+        {
+            FileManager.SaveHighScore(_user, _snake.Score);
+
+            Console.WriteLine(Logos.GameOverLogo);
+            RenderScore();
+        }
+
+        private void PopulateDirections()
```
Comment (mvlt) [inline SnakeGame/SnakeGame/Game.cs]: Why do we initialize it like this

File: SnakeGame/SnakeGame/Snake.cs [hunk 12]
```diff
@@ -0,0 +1,231 @@
...
+                if (Head.Position == snakePart.Position)
+                {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        public bool OnApple((int,int) applePosition)
```
Comment (mvlt) [inline SnakeGame/SnakeGame/Snake.cs]: Do you need this method

File: SnakeGame/SnakeGame/Snake.cs [hunk 13]
```diff
@@ -0,0 +1,231 @@
...
+                else
+                {
+                    snakeParts.Add(new SnakePart(1, i, "○"));
+                }
+            }
+
+            return snakeParts;
+        }
+
+        public void RenderSnake()
```
Comment (mvlt) [inline SnakeGame/SnakeGame/Snake.cs]: Can we handle the logic for different symbols of the snake in render method

File: SnakeGame/SnakeGame/Snake.cs [hunk 14]
```diff
@@ -0,0 +1,231 @@
...
+
+                _snakeBody[i].X = x;
+                _snakeBody[i].Y = y;
+            }
+
+            _snakeBody[0].X = Head.X;
+            _snakeBody[0].Y = Head.Y;
+        }
+
+        public bool ShouldEndGame()
```
Comment (mvlt) [inline SnakeGame/SnakeGame/Snake.cs]: Need better naming

File: SnakeGame/SnakeGame/Snake.cs [hunk 15]
```diff
@@ -0,0 +1,231 @@
...
+            {
+                Teleport();
+            }
+        }
+
+        public void UpdateBodyPosition()
+        {
+            ClearLastSnakePart();
+
+            for (int i = _snakeBody.Count - 1; i > 0; i--)
```
Comment (mvlt) [inline SnakeGame/SnakeGame/Snake.cs]: Is there better way for that job

File: SnakeGame/SnakeGame/Snake.cs [hunk 16]
```diff
@@ -0,0 +1,231 @@
...
+
+            _snakeBody[0].X = Head.X;
+            _snakeBody[0].Y = Head.Y;
+        }
+
+        public bool ShouldEndGame()
+        {
+            bool result = false;
+
+            if (CheckIfOutOfBounds())
```
Comment (mvlt) [inline SnakeGame/SnakeGame/Snake.cs]: Similar logic like shouldTeleport

File: SnakeGame/SnakeGame/Game.cs [hunk 17]
```diff
@@ -0,0 +1,184 @@
...
+            }
+
+            return hasWalls;
+        }
+
+        private void DirectionHandler(string keyPressed)
+        {
+            Directions direction;
+            
+            if (IsKeyInDictionary(keyPressed))
```
Comment (mvlt) [inline SnakeGame/SnakeGame/Game.cs]: Why wrapped in function

File: SnakeGame/SnakeGame/Game.cs [hunk 18]
```diff
@@ -0,0 +1,184 @@
...
+            {
+                direction = movementDirections[keyPressed];
+                _lastDirection = direction;
+            }
+            else
+            {
+                direction = _lastDirection;
```
Comment (mvlt) [inline SnakeGame/SnakeGame/Game.cs]: Seems a little hacky try other way

File: SnakeGame/SnakeGame/Game.cs [hunk 19]
```diff
@@ -0,0 +1,184 @@
...
+            Random random = new Random();
+
+            while (true)
+            {
+                int x = random.Next(1, Console.WindowHeight - 1);
+                int y = random.Next(1, Console.WindowWidth - 1);
+
+                _applePosition = (x, y);
+
+                if (!_snake.CheckSpawnOnSnake(_applePosition))
```
Comment (mvlt) [inline SnakeGame/SnakeGame/Game.cs]: Why wrapped

File: SnakeGame/SnakeGame/Game.cs [hunk 20]
```diff
@@ -0,0 +1,184 @@
...
+                Task game = Task.Run(() =>
+                {
+                    _keyPressed = Console.ReadKey(true).Key.ToString();
+                });
+
+                game.Wait(75);
+
+                DirectionHandler(_keyPressed);
+            }
+            while (!HasEnded);
```
Comment (mvlt) [inline SnakeGame/SnakeGame/Game.cs]: Here also checking hasEnded, so why not end here but call Over in program.cs

File: SnakeGame/SnakeGame/Snake.cs [hunk 21]
```diff
@@ -0,0 +1,231 @@
...
+
+            return false;
+        }
+
+        public bool OnApple((int,int) applePosition)
+        {
+            return applePosition == Head.Position;
+        }
+
+        public void EatApple()
```
Comment (mvlt) [inline SnakeGame/SnakeGame/Snake.cs]: Change the name to tell what the function is doing

File: SnakeGame/SnakeGame/Game.cs [hunk 22]
```diff
@@ -0,0 +1,184 @@
...
+                {"DownArrow", Directions.Down },
+                {"UpArrow", Directions.Up }
+            };
+
+        private Snake _snake = new Snake(4);
+
+        private string _user;
+        private (int, int) _applePosition;
+        private Directions _lastDirection;
+        private string _keyPressed = "RightArrow";
```
Comment (mvlt) [inline SnakeGame/SnakeGame/Game.cs]: Why do you create enum and dictionary and hardcoded here

File: SnakeGame/SnakeGame/Game.cs [hunk 23]
```diff
@@ -0,0 +1,184 @@
...
+        {
+            Right,
+            Left,
+            Up,
+            Down
+        }
+
+        public bool HasEnded = false;
+
+        private Dictionary<string, Directions> movementDirections
```
Comment (mvlt) [inline SnakeGame/SnakeGame/Game.cs]: Do we need a dictionary? Could we define the enum other way so we don't need this dictionary?

**Review by mvlt — CHANGES_REQUESTED** [review]
Think also about interfaces for some of the classes

File: SnakeGame/SnakeGame/Snake.cs [hunk 24]
```diff
@@ -0,0 +1,231 @@
...
+        {
+            WriteAt(Head.Y, Head.X, Head.Symbol);
+
+            foreach (var snakePart in _snakeBody)
+            {
+                WriteAt(snakePart.Y, snakePart.X, snakePart.Symbol);
+            }
+        }
+
+        public void ClearLastSnakePart()
```
Comment (mvlt) [inline SnakeGame/SnakeGame/Snake.cs]: I was expecting to remove it from snake body

File: SnakeGame/SnakeGame/Snake.cs [hunk 25]
```diff
@@ -0,0 +1,231 @@
...
+            bool boundY = Head.Y < 0 || Head.Y > _consoleWidth - 1;
+
+            if (boundX || boundY) { isOutOfBounds = true; }
+
+            return isOutOfBounds;
+        }
+
+        public void Teleport()
+        {
+            int x = Head.X;
```
Comment (mvlt) [inline SnakeGame/SnakeGame/Snake.cs]: Head.X is more explainable than x why assigning to a new variable

File: SnakeGame/SnakeGame/Game.cs [hunk 26]
```diff
@@ -0,0 +1,203 @@
+﻿using System.Security.Cryptography;
+
+namespace SnakeGame
+{
+    public class Game
+    {
+        public bool SnakeIsAlive = true;
+
+        private Snake _snake = new Snake(4);
```
Comment (mvlt) [inline SnakeGame/SnakeGame/Game.cs]: Better user to be able to choose starting length ex. from 0 to 10

File: SnakeGame/SnakeGame/Snake.cs [hunk 27]
```diff
@@ -0,0 +1,202 @@
...
+{
+    public class Snake
+    {
+        public ISnakePart Head { get => _body[0]; }
+
+        private List<ISnakePart> _body;
+        private (int, int) _lastBodyPartPosition;
+        private int _startingLenght;
+
+        private int _consoleHeight = Console.WindowHeight;
```
Comment (mvlt) [inline SnakeGame/SnakeGame/Snake.cs]: why do we need to store height and width in fields

File: SnakeGame/SnakeGame/Snake.cs [hunk 28]
```diff
@@ -0,0 +1,202 @@
...
+
+            for (int i = snakeLenght; i >= 1; i--)
+            {
+                snakeParts.Add(new SnakePart(1, i));
+            }
+
+            return snakeParts;
+        }
+
+        private bool MoveIsNotPossible(bool hasWalls)
```
Comment (mvlt) [inline SnakeGame/SnakeGame/Snake.cs]: Not the best name for this function

File: SnakeGame/SnakeGame/Game.cs [hunk 29]
```diff
@@ -0,0 +1,203 @@
...
+
+            OpenStartingMenu();
+            // Close starting menu
+            Console.Clear();
+
+            StartingProcesses();
+
+            GameLoop();
+
+            if (!SnakeIsAlive)
```
Comment (mvlt) [inline SnakeGame/SnakeGame/Game.cs]: Do we need to check if Snake is Alive here

File: SnakeGame/SnakeGame/Game.cs [hunk 30]
```diff
@@ -0,0 +1,203 @@
...
+                if (!_snake.CheckSpawnOnBody(_applePosition))
+                {
+                    Console.SetCursorPosition(y, x);
+                    Console.Write("@");
+                    break;
+                }
+            }
+        }
+
+        private void GameOver()
```
Comment (mvlt) [inline SnakeGame/SnakeGame/Game.cs]: Game over not best name if also is saving the score

File: SnakeGame/SnakeGame/Game.cs [hunk 31]
```diff
@@ -0,0 +1,203 @@
...
+        {
+            FileManager.SaveHighScore(_user, _snake.Score);
+
+            Console.WriteLine(Logos.GameOverLogo);
+            RenderScore();
+        }
+
+        private void AppleHandler()
```
Comment (mvlt) [inline SnakeGame/SnakeGame/Game.cs]: AppleHandler is also adding snake parts

File: SnakeGame/SnakeGame/Game.cs [hunk 32]
```diff
@@ -0,0 +1,203 @@
...
+
+        public Game() { }
+
+        public void MainProcess()
+        {
+            Console.CursorVisible = false;
+
+            OpenStartingMenu();
+            // Close starting menu
+            Console.Clear();
```
Comment (mvlt) [inline SnakeGame/SnakeGame/Game.cs]: Can we clear the console in some of the functions which are called in the Main Process

File: SnakeGame/SnakeGame/Game.cs [hunk 33]
```diff
@@ -0,0 +1,203 @@
...
+                    break;
+                }
+
+                SetLegalDirection();
+
+                DirectionHandler();
+            }
+        }
+
+        private void StartingProcesses()
```
Comment (mvlt) [inline SnakeGame/SnakeGame/Game.cs]: Why do we need to spawn apple and render the snake here

File: SnakeGame/SnakeGame/Game.cs [hunk 34]
```diff
@@ -0,0 +1,203 @@
...
+            }
+        }
+
+        private void GetConfigData()
+        {
+            SetUserName();
+            _hasWalls = AskHasWalls();
+        }
+
+        private void WaitForKeyPress()
```
Comment (mvlt) [inline SnakeGame/SnakeGame/Game.cs]: Waiting for key press is also writing starting logo

File: SnakeGame/SnakeGame/Snake.cs [hunk 35]
```diff
@@ -0,0 +1,202 @@
...
+            }
+            else if (CheckIfHitItself())
+            {
+                result = true;
+            }
+
+            return result;
+        }
+
+        private bool CheckTeleport(bool hasWalls)
```
Comment (mvlt) [inline SnakeGame/SnakeGame/Snake.cs]: better name

**Review by mvlt — CHANGES_REQUESTED** [review]

File: SnakeGame/SnakeGame/Classes/Game.cs [hunk 36]
```diff
@@ -0,0 +1,219 @@
...
+using SnakeGame.Utils;
+
+namespace SnakeGame.Classes
+{
+    public class Game
+    {
+        public bool SnakeIsAlive = true;
+
+        private Snake _snake;
+        private IDiffilcultyHandler _diffilcultyHandler = new DiffilcultyHandler();
```
Comment (mvlt) [inline SnakeGame/SnakeGame/Classes/Game.cs]: Syntax error. Violation of Dependency Inversion Principle.

File: SnakeGame/SnakeGame/Classes/Game.cs [hunk 37]
```diff
@@ -0,0 +1,219 @@
...
+        private IScoreManager _scoreManager = new ScoreManager();
+
+        private string _user;
```
Comment (mvlt) [inline SnakeGame/SnakeGame/Classes/Game.cs]: Violation of Single Responsibility Principle.

File: SnakeGame/SnakeGame/Classes/Game.cs [hunk 38]
```diff
@@ -0,0 +1,219 @@
...
+                    break;
+                }
+
+                SetLegalDirection();
+
+                DirectionHandler();
+            }
+        }
+
+        private void OpenStartingMenu()
```
Comment (mvlt) [inline SnakeGame/SnakeGame/Classes/Game.cs]: SRP

File: SnakeGame/SnakeGame/Classes/Game.cs [hunk 39]
```diff
@@ -0,0 +1,219 @@
...
+
+        private void SetSnakeLength()
+        {
+            Console.WriteLine("How long would you like the snake to be?");
+            Console.WriteLine("Minimum of 0, Maximum of 10");
+            _snakeLength = int.Parse(Console.ReadLine());
+
+            if (_snakeLength < 0 || _snakeLength > 10)
+            {
+                throw new ArgumentOutOfRangeException
```
Comment (mvlt) [inline SnakeGame/SnakeGame/Classes/Game.cs]: SRP

File: SnakeGame/SnakeGame/Classes/Game.cs [hunk 40]
```diff
@@ -0,0 +1,219 @@
...
+                Thread.Sleep(250);
+        }
+
+        private static void PrintStartingLogo()
+        {
+            Console.Clear();
+            Console.WriteLine(Logos.GameStartLogo);
+        }
+
+        private void SpawnApple()
```
Comment (mvlt) [inline SnakeGame/SnakeGame/Classes/Game.cs]: SRP

File: SnakeGame/SnakeGame/Classes/Game.cs [hunk 41]
```diff
@@ -0,0 +1,219 @@
...
+
+        private void GameOverScreen()
+        {
+            Console.WriteLine(Logos.GameOverLogo);
+            _scoreManager.RenderLogo();
+        }
+
+        private bool IsOnApple()
+        {
+            return _applePosition == _snake.Head.Position;
```
Comment (mvlt) [inline SnakeGame/SnakeGame/Classes/Game.cs]: Is this responsibility for the game or the snake

**Review by mvlt — CHANGES_REQUESTED** [review]

File: SnakeGame/SnakeGame/Classes/Snake.cs [hunk 42]
```diff
@@ -0,0 +1,191 @@
...
+
+            for (int i = snakeLenght; i >= 0; i--)
+            {
+                snakeParts.Add(new SnakePart(1, i));
+            }
+
+            return snakeParts;
+        }
+
+        private bool MoveIsIllegal(bool hasWalls)
```
Comment (mvlt) [inline SnakeGame/SnakeGame/Classes/Snake.cs]: This name sounds more like a statement than a function

File: SnakeGame/SnakeGame/Classes/Snake.cs [hunk 43]
```diff
@@ -0,0 +1,191 @@
...
+            (int x, int y) = _lastBodyPartPosition;
+            _body.Add(new SnakePart(x, y));
+        }
+
+        public bool CheckSpawnOnBody((int, int) applePosition)
+        {
+            return _body.Any(x => x.Position == applePosition);
+        }
+
+        public void MoveX(int x, bool gameHasWalls)
```
Comment (mvlt) [inline SnakeGame/SnakeGame/Classes/Snake.cs]: x ?

File: SnakeGame/SnakeGame/Classes/Game.cs [hunk 44]
```diff
@@ -0,0 +1,149 @@
...
+using SnakeGame.Interfaces;
+using SnakeGame.Utils;
+
+namespace SnakeGame.Classes
+{
+    public class Game : IGame
+    {
+        public bool snakeIsAlive = true;
+
+        private IScoreManager _scoreManager =
```
Comment (mvlt) [inline SnakeGame/SnakeGame/Classes/Game.cs]: We could pass scoreManager as a parameter to the Game class and initialize it in the constructor

File: SnakeGame/SnakeGame/Classes/Game.cs [hunk 45]
```diff
@@ -0,0 +1,149 @@
...
+            Factory.CreateScoreManager();
+
+        private IDiffilcultyHandler _diffilcultyHandler =
```
Comment (mvlt) [inline SnakeGame/SnakeGame/Classes/Game.cs]: DifficultyHandler also

File: SnakeGame/SnakeGame/Classes/Game.cs [hunk 46]
```diff
@@ -0,0 +1,149 @@
...
+            _config = gameConfig;
+        }
+
+        public void MainProcess()
+        {
+            Console.CursorVisible = false;
+
+            Menu.StartingScreen();
+
+            _snake = Factory.CreateSnake(_config.SnakeLength);
```
Comment (mvlt) [inline SnakeGame/SnakeGame/Classes/Game.cs]: Snake could also be passed to the constructor

File: SnakeGame/SnakeGame/Classes/Snake.cs [hunk 47]
```diff
@@ -0,0 +1,156 @@
...
+    public class Snake : ISnake
+    {
+        private (int, int) _lastBodyPartPosition;
+        private IPoint _newPosition =
+            Factory.CreatePoint(0, 0);
+
+        public IPoint Head { get => Body[0]; }
+        public List<IPoint> Body { get; private set; }
+
+        public (int, int) CurrentPosition
```
Comment (mvlt) [inline SnakeGame/SnakeGame/Classes/Snake.cs]: Are we using this only in the tests

File: SnakeGame/SnakeGame/Classes/SnakeMoveChecker.cs [hunk 48]
```diff
@@ -0,0 +1,54 @@
...
+{
+    internal class SnakeMoveChecker
+    {
+        public static bool IsValid(
+            List<IPoint> snakeBody, IPoint newPosition, bool gameHasWalls)
+        {
+            bool isValid = true;
+
+            // добави exception class
+            if (CheckIfOutOfBounds(newPosition) && !gameHasWalls)
```
Comment (mvlt) [inline SnakeGame/SnakeGame/Classes/SnakeMoveChecker.cs]: If game has no walls no need to evaluate CheckIfOutOfBounds

File: SnakeGame/SnakeGame/Utils/ScoreManager.cs [hunk 49]
```diff
@@ -0,0 +1,36 @@
...
+        {
+            Score += n;
+        }
+
+        public void Render()
+        {
+            Writer.WriteAt(0, 0, $"Score: {Score}");
+        }
+
+        public void RenderLogo()
```
Comment (mvlt) [inline SnakeGame/SnakeGame/Utils/ScoreManager.cs]: We could have something like fancyMessageRenderer which takes the score and do the rendering of all logos. I think score manager should not be responsible for that also

File: SnakeGame/SnakeGame/Utils/StandardMessages.cs [hunk 50]
```diff
@@ -0,0 +1,37 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace SnakeGame.Utils
+{
+    public class StandardMessages
```
Comment (mvlt) [inline SnakeGame/SnakeGame/Utils/StandardMessages.cs:9]: We are not only storing the message content but we also writing them better name for the class

**Review by mvlt — CHANGES_REQUESTED** [review]

File: SnakeGame/SnakeGame/Classes/Food.cs [hunk 51]
```diff
@@ -0,0 +1,42 @@
...
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using System.Xml.Linq;
+
+namespace SnakeGame.Classes
+{
+    public class Food : Point, IFood
```
Comment (mvlt) [inline SnakeGame/SnakeGame/Classes/Food.cs]: Renderable?

File: SnakeGame/SnakeGame/Classes/Game.cs [hunk 52]
```diff
@@ -0,0 +1,136 @@
...
+            _foodHandler = foodHandler;
+            _scoreManager = scoreManager;
+            _snake = snake;
+        }
+
+        public void MainProcess()
+        {
+            Console.CursorVisible = false;
+
+            ScreenRenderer.StartingScreen();
```
Comment (mvlt) [inline SnakeGame/SnakeGame/Classes/Game.cs]: StartingScreen no verb in the name of the function

File: SnakeGame/SnakeGame/Classes/Game.cs [hunk 53]
```diff
@@ -0,0 +1,136 @@
...
+
+            GameLoop();
+
+            Writer.FileWrite(GameConfig.Player, _scoreManager.CurrentScore);
+            ScreenRenderer.GameOverScreen(_scoreManager);
```
Comment (mvlt) [inline SnakeGame/SnakeGame/Classes/Game.cs]: No verb in GameOverScreen and why need the dependence to scoreManager

File: SnakeGame/SnakeGame/Classes/Game.cs [hunk 54]
```diff
@@ -0,0 +1,136 @@
...
+                OnAction();
+            }
+
+            DisableTimers();
+        }
+
+        private void EnableTimers()
+        {
+            _timers.Add(new Timer((e)
+                => TimerCallback(Factory.CreateSwitch), null, 10000, 10000));
```
Comment (mvlt) [inline SnakeGame/SnakeGame/Classes/Game.cs]: Magic numbers

File: SnakeGame/SnakeGame/Utils/ScoreManager.cs [hunk 55]
```diff
@@ -0,0 +1,42 @@
...
+            CurrentScore = (snakeBody.Count - 1) - GameConfig.InitalSnakeLength;
+            diffilcultyHandler.CheckToDecreaseLevel(CurrentScore, PreviousScore);
+        }
+
+        public void Render()
+        {
+            Writer.ConsoleWriteAt(0, 0, $"Score: {CurrentScore} ");
+        }
+
+        public void CheckScoreUnderZero(ISnake snake)
```
Comment (mvlt) [inline SnakeGame/SnakeGame/Utils/ScoreManager.cs]: Do we need scoreManager to be dependent of the snake?

File: SnakeGame/SnakeGame/Utils/ScreenRenderer.cs [hunk 56]
```diff
@@ -0,0 +1,41 @@
...
+        public static void StartingScreen()
+        {
+            PrintStartingLogo();
+            WaitForKeyPress();
+            Console.Clear();
+        }
+
+        public static void GameOverScreen(IScoreManager scoreManager)
+        {
+            Console.WriteLine(Logos.GameOverLogo);
```
Comment (mvlt) [inline SnakeGame/SnakeGame/Utils/ScreenRenderer.cs]: Why not using the writer?

File: SnakeGame/SnakeGame/Utils/Validator.cs [hunk 57]
```diff
@@ -0,0 +1,22 @@
+﻿using SnakeGame.Interfaces;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace SnakeGame.Utils
+{
+    public class Validator
```
Comment (mvlt) [inline SnakeGame/SnakeGame/Utils/Validator.cs]: This seems more like snakeLengthValidator

File: SnakeGame/SnakeGame/Classes/Game.cs [hunk 58]
```diff
@@ -0,0 +1,136 @@
...
+
+                _snake.Render();
+
+                if (_foodHandler.SnakeOnFood(_snake.Head))
+                {
+                    _foodHandler.Handle(
+                        _snake,
+                        _scoreManager);
+
+                    _scoreManager.Set(_snake.Body, _diffilcultyHandler);
```
Comment (mvlt) [inline SnakeGame/SnakeGame/Classes/Game.cs]: Why scoreManager also need to know about defficultyHandler

File: SnakeGame/SnakeGame/Classes/Snake.cs [hunk 59]
```diff
@@ -0,0 +1,182 @@
...
+            }
+        }
+
+        public void AddPart()
+        {
+            (int x, int y) = _lastBodyPart.Position;
+            Body.Add(Factory.CreatePoint(x, y));
+        }
+
+        public void RemovePart()
```
Comment (mvlt) [inline SnakeGame/SnakeGame/Classes/Snake.cs]: Remove tail better. With name RemovePrat I have the feeling that with this function I can remove every part but I can't

File: SnakeGame/SnakeGame/Classes/Snake.cs [hunk 60]
```diff
@@ -0,0 +1,182 @@
...
+        {
+            ClearLastBodyPart();
```
Comment (mvlt) [inline SnakeGame/SnakeGame/Classes/Snake.cs]: RemovePart is calling ClearLastBodyPart with this naming I am assuming they are doing same thing

File: SnakeGame/SnakeGame/Classes/Snake.cs [hunk 61]
```diff
@@ -0,0 +1,182 @@
...
+            Body.RemoveAt(Body.Count - 1);
+        }
+
+        public void Move()
+        {
+            switch (Direction)
+            {
+                case Directions.RightArrow:
+                    MoveY(1);
```
Comment (mvlt) [inline SnakeGame/SnakeGame/Classes/Snake.cs]: Magic number

File: SnakeGame/SnakeGame/Classes/Snake.cs [hunk 62]
```diff
@@ -0,0 +1,182 @@
...
+            }
+        }
+
+        private void MoveX(int value)
+        {
+            _newPosition.Position = (Head.X + value, Head.Y);
+
+            var isValid = SnakeMoveChecker.Check(this, _newPosition);
+
+            if (isValid == true)
```
Comment (mvlt) [inline SnakeGame/SnakeGame/Classes/Snake.cs]: Boolean expression with three cases?

File: SnakeGame/SnakeGame/Utils/SnakeMoveChecker.cs [hunk 63]
```diff
@@ -0,0 +1,82 @@
...
+using SnakeGame.Constants;
+using SnakeGame.Interfaces;
+using System.Xml.Linq;
+
+namespace SnakeGame.Utils
+{
+    internal class SnakeMoveChecker
+    {
+        public static bool? Check(
+            ISnake snake, IPoint newPosition)
```
Comment (mvlt) [inline SnakeGame/SnakeGame/Utils/SnakeMoveChecker.cs]: Do we need the whole snake object

File: SnakeGame/SnakeGame/Utils/Writer.cs [hunk 64]
```diff
@@ -0,0 +1,44 @@
...
+
+        private static string _filePath
+            = @"C:\Users\todor.zagorov\source\repos\c-sharp-oop\SnakeGame\SnakeGame\HighScores.txt";
+
+        public static void ConsoleWriteAt(int y, int x, string str)
+        {
+            while (true)
+            {
+                if (_isRendering == false)
+                {
```
Comment (mvlt) [inline SnakeGame/SnakeGame/Utils/Writer.cs:24]: This logic seems strange. Is there more simple approach?

File: SnakeGame/SnakeGame/Interfaces/IOnConsume.cs [hunk 65]
```diff
@@ -0,0 +1,14 @@
+﻿using SnakeGame.Classes;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace SnakeGame.Interfaces
+{
+    public interface IOnConsume
```
Comment (mvlt) [inline SnakeGame/SnakeGame/Interfaces/IOnConsume.cs]: Consumable?

**Review by mvlt — COMMENTED** [review]
Do we need all implemented interfaces

File: SnakeGame/SnakeGame/Classes/Game.cs [hunk 66]
```diff
@@ -0,0 +1,136 @@
...
+            _foodHandler = foodHandler;
+            _scoreManager = scoreManager;
+            _snake = snake;
+        }
+
+        public void MainProcess()
+        {
+            Console.CursorVisible = false;
+
+            ScreenRenderer.StartingScreen();
```
Comment (mvlt) [inline SnakeGame/SnakeGame/Classes/Game.cs]: Again no verb in the name of the function

File: SnakeGame/SnakeGame/Classes/Game.cs [hunk 67]
```diff
@@ -0,0 +1,136 @@
...
+
+            GameLoop();
+
+            Writer.FileWrite(GameConfig.Player, _scoreManager.CurrentScore);
+            ScreenRenderer.GameOverScreen(_scoreManager);
```
Comment (mvlt) [inline SnakeGame/SnakeGame/Classes/Game.cs]: And here why marked as resolved

File: SnakeGame/SnakeGame/Classes/Strategies/AddSnakePartStrategy.cs [hunk 68]
```diff
@@ -0,0 +1,21 @@
...
+using SnakeGame.Utils;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace SnakeGame.Classes.Strategies
+{
+    public class AddSnakePartStrategy : IConsumable
```
Comment (mvlt) [inline SnakeGame/SnakeGame/Classes/Strategies/AddSnakePartStrategy.cs:11]: How many times this strategy will be instantiated?

File: SnakeGame/SnakeGame/Constants/Logos.cs [hunk 69]
```diff
@@ -0,0 +1,122 @@
...
+
+        public static string ScoreWordLogo
+        {
+            get =>
+            "\r\n█▀▀▀█ █▀▀ █▀▀█ █▀▀█ █▀▀ ▄" +
+            "\r\n▀▀▀▄▄ █   █  █ █▄▄▀ █▀▀  " +
+            "\r\n█▄▄▄█ ▀▀▀ ▀▀▀▀ ▀ ▀▀ ▀▀▀ ▀";
+        }
+
+        public static string GenerateScoreLogo(int scoreInt)
```
Comment (mvlt) [inline SnakeGame/SnakeGame/Constants/Logos.cs]: Should this functionality be part of Logos class

**Review by mvlt — CHANGES_REQUESTED** [review]

File: SnakeGame/SnakeGame/Classes/Game.cs (hunk 67 above)
Comment (todorzag) [reply SnakeGame/SnakeGame/Classes/Game.cs]: Sorry my bad, confusion on my part because of Menu.cs

**Pull request merged**